cdap-master/src/main/java/co/cask/cdap/data/tools/MetricsKafkaUpgrader.java

```diff
@@ -0,0 +1,167 @@
+import org.apache.hadoop.hbase.client.HTable;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.twill.filesystem.LocationFactory;
+import org.apache.twill.kafka.client.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Upgrade kafka metrics meta table data
+ */
+public class MetricsKafkaUpgrader extends AbstractUpgrader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DatasetUpgrader.class);
+  private static final byte[] OFFSET_COLUMN = Bytes.toBytes("o");
+
+  private final CConfiguration cConf;
+  private final Configuration hConf;
+  private final HBaseTableUtil hBaseTableUtil;
+  private final DatasetFramework dsFramework;
+  private final String kafkaTableName;
+
+  @Inject
+  public MetricsKafkaUpgrader(CConfiguration cConf, Configuration hConf, LocationFactory locationFactory,
+                              HBaseTableUtil hBaseTableUtil, final DatasetFramework dsFramework) {
+    super(locationFactory);
+    this.cConf = cConf;
+    this.hConf = hConf;
+    this.hBaseTableUtil = hBaseTableUtil;
+    this.dsFramework = dsFramework;
+    this.kafkaTableName =  cConf.get(MetricsConstants.ConfigKeys.KAFKA_META_TABLE,
+                                     MetricsConstants.DEFAULT_KAFKA_META_TABLE);
+  }
+
+  private MetricsTable getOrCreateKafkaTable(String tableName, DatasetProperties props) {
+    MetricsTable table = null;
+    // old kafka table is in the default namespace
+    Id.DatasetInstance metricsDatasetInstanceId = Id.DatasetInstance.from(Constants.DEFAULT_NAMESPACE_ID, tableName);
+    try {
+      table = DatasetsUtil.getOrCreateDataset(dsFramework, metricsDatasetInstanceId,
+                                              MetricsTable.class.getName(), props, null, null);
+    } catch (Exception e) {
+      LOG.error("Exception while creating table {}.", tableName, e);
+    }
+    return table;
+  }
+
+  @Override
+  public void upgrade() throws Exception {
+    KafkaConsumerMetaTable kafkaMetaTableDestination =
+      new DefaultMetricDatasetFactory(cConf, dsFramework).createKafkaConsumerMeta();
+    // copy kafka offset from old table to new kafka metrics tabl
+    String kafkaTableNameOld = Joiner.on(".").join(Constants.SYSTEM_NAMESPACE, "default", kafkaTableName);
+    try {
+      // assuming we are migrating from 2.6 or 2.7
+      HTable hTable = getHTable(kafkaTableNameOld);
+      byte[] columnFamily = getColumnFamily(kafkaTableNameOld);
+
+      // iterate old table and copy all rows to the new kafkaConsumerMetaTable
+      try {
+        LOG.info("Starting upgrade for table {}", Bytes.toString(hTable.getTableName()));
+        Scan scan = getScan(columnFamily);
+        ResultScanner resultScanner = hTable.getScanner(scan);
+        Result result;
+        try {
```

Comment:
Swap this and the line before. It's generally a good practice you have the 'try' begin immediately after the resource creation that you need to free in finally.